R/derive_param_confirmed_resp.R

```diff
@@ -0,0 +1,386 @@
+  assert_s3_class(source_pd, "date_source", optional = TRUE)
+  assert_list_of(source_datasets, "data.frame")
+  assert_integer_scalar(ref_confirm, subset = "non-negative")
+  assert_integer_scalar(max_nr_ne, subset = "non-negative")
+  assert_logical_scalar(accept_sd)
+  assert_function(aval_fun)
+  assert_varval_list(set_values_to, required_elements = "PARAMCD")
+  assert_vars(subject_keys)
+  assert_data_frame(
+    dataset,
+    required_vars = quo_c(subject_keys, vars(PARAMCD, ADT, AVALC))
+  )
+  assert_data_frame(dataset_adsl, required_vars = subject_keys)
+  if (!is.null(dataset)) {
+    assert_param_does_not_exist(dataset, quo_get_expr(set_values_to$PARAMCD))
+  }
+
+  # Restrict input dataset
+  source_data <- dataset %>%
+    filter_pd(
+      filter = !!filter_source,
+      source_pd = source_pd,
+      source_datasets = source_datasets
+    )
+
+  # Check for invalid AVALC values
+  resp_vals <- source_data$AVALC
+  valid_vals <- c("CR", "PR", "SD", "NON-CR/NON-PD", "PD", "NE", "ND")
+  invalid_vals <- unique(resp_vals[!resp_vals %in% valid_vals])
+  if (length(invalid_vals) > 0) {
+    abort(
+      paste0(
+        "The function is considering only the following response values:\n",
+        enumerate(valid_vals),
+        "\nThe following invalid values were found:\n",
+        enumerate(invalid_vals)
+      )
+    )
+  }
+
+  # Create observations for potential responses
+  cr_data <- filter_confirmation(
+    source_data,
+    by_vars = subject_keys,
+    join_vars = vars(AVALC, ADT),
+    order = vars(ADT),
+    first_cond = AVALC.join == "CR" &
+      ADT.join >= ADT + days(ref_confirm),
+    filter = AVALC == "CR" &
+      all(AVALC.join %in% c("CR", "NE")) &
+      count_vals(var = AVALC.join, val = "NE") <= max_nr_ne
+  ) %>%
+    mutate(
+      AVALC = "Y",
+      tmp_order = 1
+    )
+
+  if (accept_sd) {
+    max_nr_sd <- 1
+  } else {
+    max_nr_sd <- 0
+  }
+  pr_data <- filter_confirmation(
+    source_data,
+    by_vars = subject_keys,
+    join_vars = vars(AVALC, ADT),
+    order = vars(ADT),
+    first_cond = AVALC.join %in% c("CR", "PR") &
+      ADT.join >= ADT + days(ref_confirm),
+    filter = AVALC == "PR" &
+      all(AVALC.join %in% c("CR", "PR", "SD", "NE")) &
+      count_vals(var = AVALC.join, val = "NE") <= max_nr_ne &
+      count_vals(var = AVALC.join, val = "SD") <= max_nr_sd &
+      (
+        min_cond(
+          var = ADT.join,
+          cond = AVALC.join == "CR"
+        ) > max_cond(var = ADT.join, cond = AVALC.join == "PR") |
+          count_vals(var = AVALC.join, val = "CR") == 0
+      )
+  ) %>%
+    mutate(
+      AVALC = "Y",
+      tmp_order = 1
+    )
+
+  source_vars <- colnames(source_data)
+  adsl_vars <- colnames(dataset_adsl)
+
+  missing_data <- dataset_adsl %>%
+    select(!!!subject_keys, intersect(source_vars, adsl_vars)) %>%
+    mutate(
+      AVALC = "N",
+      tmp_order = 2
+    )
+
+  # Select response
+  rsp <- bind_rows(cr_data, pr_data, missing_data) %>%
```

Comment:
In extreme case if the subject qualifies for confirmed CR and also PR. As such for objective response it should not matter, but for actual BOR it matters,  I see you order using using tmp_order and adt , which make sense for objective response Yes/No but for BOR we should use CR in that case? any thoughts?

Reply:
We do not need `tmp_order` here because NAs are sorted to the end. I.e., if we just order by `ADT`, an observation from `missing_data` is selected only if there is no observation in `cr_data` or `pr_data`.

I'll update the code.

Reply:
thanks